src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Query/PropertyPages/LaunchProfileFromProjectDataProducer.cs

```diff
@@ -0,0 +1,119 @@
+﻿// Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE.md file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Build.Framework.XamlTypes;
+using Microsoft.VisualStudio.ProjectSystem.Debug;
+using Microsoft.VisualStudio.ProjectSystem.Properties;
+using Microsoft.VisualStudio.ProjectSystem.Query;
+using Microsoft.VisualStudio.ProjectSystem.Query.Frameworks;
+using Microsoft.VisualStudio.ProjectSystem.Query.ProjectModel;
+using Microsoft.VisualStudio.ProjectSystem.Query.ProjectModel.Implementation;
+using Microsoft.VisualStudio.ProjectSystem.VS.Utilities;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Query.PropertyPages
+{
+    internal class LaunchProfileFromProjectDataProducer : QueryDataFromProviderStateProducerBase<UnconfiguredProject>
+    {
+        private readonly ILaunchProfilePropertiesAvailableStatus _properties;
+        private readonly IPropertyPageQueryCacheProvider _queryCacheProvider;
+
+        public LaunchProfileFromProjectDataProducer(ILaunchProfilePropertiesAvailableStatus properties, IPropertyPageQueryCacheProvider queryCacheProvider)
+        {
+            _properties = properties;
+            _queryCacheProvider = queryCacheProvider;
+        }
+
+        protected override Task<IEnumerable<IEntityValue>> CreateValuesAsync(IEntityValue parent, UnconfiguredProject providerState)
+        {
+            return CreateLaunchProfileValuesAsync(parent, providerState);
+        }
+
+        private async Task<IEnumerable<IEntityValue>> CreateLaunchProfileValuesAsync(IEntityValue parent, UnconfiguredProject project)
+        {
+            if (project.Services.ExportProvider.GetExportedValueOrDefault<ILaunchSettingsProvider>() is ILaunchSettingsProvider launchSettingsProvider
+                && await project.GetProjectLevelPropertyPagesCatalogAsync() is IPropertyPagesCatalog projectCatalog
+                && await launchSettingsProvider.WaitForFirstSnapshot(Timeout.Infinite) is ILaunchSettings launchSettings)
+            {
+                return createLaunchProfileValues();
+            }
+
+            return Enumerable.Empty<IEntityValue>();
+
+            IEnumerable<IEntityValue> createLaunchProfileValues()
```

Comment:
Instead of creating this local function, it might make sense to have the method return `IAsyncEnumerable` instead, so you can still `yield return` the results you want. Info here: https://agirlamonggeeks.com/2018/11/25/c-8-features-part-2-async-method-with-yield-return/